### PR TITLE
[Data] Add `darwin`, `x86` constraints for `pyarrow<18` restriction

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?

Addresses the issue in https://github.com/ray-project/ray/pull/48300/files#r1821887632.  Make
sure to add darwin and x86 constraint in all necessary places.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by can-anyscale — APPROVED** [review]

**Comment by can-anyscale** [comment]
To get this done done, do we need to update these test jobs https://buildkite.com/ray-project/postmerge/builds/6755#0192dbd2-44c7-4ade-b520-673a02c2d5fe to use pyarrow 18 as well?

**Comment by scottjlee** [comment]
@can-anyscale those will explicitly install pyarrow 6/17/nightly versions, right? since we are not officially supporting pyarrow 18 with ray data yet, i don't think we need to do it until then.

**Comment by can-anyscale** [comment]
@scottjlee 100% yes; but with this change `pip install ray[data]` will install pyarrow 18 by default though so it just feels weird that the default behavior is not supported, tested, etc. Not a blocker, just something to call out for.

File: python/setup.py [hunk 1]
```diff
@@ -227,7 +227,9 @@ def get_packages(self):
 if setup_spec.type == SetupType.RAY:
     pandas_dep = "pandas >= 1.3"
     numpy_dep = "numpy >= 1.20"
-    pyarrow_dep = "pyarrow >= 6.0.1,<18"
+    pyarrow_dep = (
+        "pyarrow >= 6.0.1,<18" if sys.platform == "darwin" else "pyarrow >= 6.0.1"
+    )
```
Comment (aslonnie) [inline python/setup.py:232]: this should be in the python deps constraint format (not in python conditions).

you can use line 235 as an example: `"cupy-cuda12x; sys_platform != 'darwin'"`

**Comment by scottjlee** [comment]
> @scottjlee 100% yes; but with this change `pip install ray[data]` will install pyarrow 18 by default though so it just feels weird that the default behavior is not supported, tested, etc. Not a blocker, just something to call out for.

Ah, i understand you now. Let me work with the team to try adding support for pyarrow 18. if tests don't break, it should be a smooth process to enable it

File: python/setup.py [hunk 2]
```diff
@@ -227,7 +227,10 @@ def get_packages(self):
 if setup_spec.type == SetupType.RAY:
     pandas_dep = "pandas >= 1.3"
     numpy_dep = "numpy >= 1.20"
-    pyarrow_dep = "pyarrow >= 6.0.1,<18"
+    pyarrow_deps = [
+        "pyarrow >= 6.0.1,<18; sys_platform == 'darwin'",
```
Comment (aslonnie) [inline python/setup.py:231]: just for our future selves, pyarrow 18 starts releasing for macos 12, where ray is still releasing for macos 10.

we are capping as we are even already thinking about stop supporting intel silicon.. the world of macos is moving to apple silicon.

**Review by aslonnie — APPROVED** [review]

File: python/setup.py (hunk 2 above)
Comment (aslonnie) [reply python/setup.py:231]: > pyarrow 18 starts releasing for macos 12, where ray is still releasing for macos 10.

thanks @can-anyscale for discovering this.

Comment (aslonnie) [reply python/setup.py:231]: not sure, but maybe limit to x86 too? I think the issue that blocks the release tests is intel silicon only.

Comment (aslonnie) [reply python/setup.py:231]: > but maybe limit to x86 too

I am also kind of okay with its current form.. but limiting `<18` to x86 arch feels better. I am okay with just firing all the intel silicon users, but I am a bit worry that some macos apple silicon ray user in the future want to use with pyarrow 18+.

Comment (scottjlee) [reply python/setup.py:231]: > limiting <18 to x86 arch feels better

something like this @aslonnie ?

```
pyarrow_deps = [
        "pyarrow >= 6.0.1,<18; sys_platform == 'darwin' and platform_machine == 'x86_64'",
        "pyarrow >= 6.0.1; sys_platform != 'darwin' or platform_machine != 'x86_64'",
    ]
```

Comment (aslonnie) [reply python/setup.py:231]: yeah exactly.

I personally would write:

````
pyarrow >= 6.0.1
pyarrow <18; sys_platform == 'darwin' and platform_machine == 'x86_64'
````

Comment (aslonnie) [reply python/setup.py:231]: logically equivalent, but might reads clearer on the intention.

Comment (scottjlee) [reply python/setup.py:231]: much cleaner, thanks!

Comment (aslonnie) [reply python/setup.py:231]: don't forget to sync to the other place (`python/requirements.txt`) too.

**Review by aslonnie — APPROVED** [review]

**Pull request merged**